documents/api/filtersets.py

```diff
@@ -31,16 +29,30 @@ def filter(self, qs, value):
 
 class MetadataJSONFilter(Filter):
     """
-    Custom filter for free searching from metadata field.
-    Cast json to text and annotate it to queryset for filtering
+    Custom filter for metadata field.
+    Query can be made with key value pairs like "key:value, ..." without quotes
     """
 
     def filter(self, qs, value):
         if value in EMPTY_VALUES:
             return qs
-        return qs.annotate(metatext=Cast("metadata", TextField())).filter(
-            metatext__contains=value
-        )
+        query_parts = [val.strip() for val in value.split(",")]
+        for part in query_parts:
+            try:
+                key, value = part.split(":", maxsplit=1)
+                qs = qs.filter(
+                    metadata__has_key=key, **{"metadata__" + key + "__icontains": value}
+                )
+            except ValueError:
+                raise ValidationError(
```

Comment:
It seems bit harsh that filter raises exception. But maybe it should work like this for unknown reasons.

Reply:
This custom filter requires the query to be in certain format for it to work. Exception is raised to inform user how the filter is used properly.